linera-core/src/client/mod.rs

```diff
@@ -1132,76 +1180,80 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut certificates: Vec<Certificate> = Vec::new();
-        let mut new_tracker = tracker;
+        let mut tracker = MultichainTracker::new(tracker, info.requested_received_log.clone());
+        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
+            BTreeMap::<ChainId, Vec<BlockHeight>>::new(),
+            |mut chain_to_info, entry| {
+                chain_to_info
+                    .entry(entry.chain_id)
+                    .or_default()
+                    .push(entry.height);
+                chain_to_info
+            },
+        );
 
-        for entry in info.requested_received_log {
-            let query = ChainInfoQuery::new(entry.chain_id)
-                .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+        // Collect all certificates that have been successfully processed.
+        let mut certificates: Vec<Certificate> = Vec::new();
 
-            let local_response = self
-                .client
-                .local_node
-                .handle_chain_info_query(query.clone())
-                .await
-                .map_err(|error| NodeError::LocalNodeQuery {
-                    error: error.to_string(),
-                })?
-                .info;
-            if !local_response.requested_sent_certificate_hashes.is_empty() {
-                // We've already processed incoming messages for this certificate.
-                new_tracker += 1;
-                continue;
-            }
+        for (chain_id, mut block_batch) in remote_node_chains_view {
+            block_batch.sort();
+            let first_block = block_batch.first().expect("nonempty batch");
+            let batch_size = block_batch.len() as u64;
```

Comment:
The heights usually have gaps: Not every block in the sending chain may have sent a message to us.

So we should really use `block_batch.last().0 - first_block.0 + 1` as the size.

Reply:
`block_batch` isn't necessarily contiguous so your formula doesn't work either, does it? Isn't it equivalent to `.len()` 👀 The batch size here isn't about the messages sent but about the blocks, right?

Reply:
Done. Thanks for catching.